tests/OmniSharp.MSBuild.Tests/OnDemandProjectsLoadTests.cs

```diff
@@ -0,0 +1,162 @@
+using OmniSharp.MSBuild.Models;
+using OmniSharp.Options;
+using OmniSharp.Roslyn.CSharp.Services.Refactoring.V2;
+using OmniSharp.Roslyn.CSharp.Services.Structure;
+using TestUtility;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace OmniSharp.MSBuild.Tests
+{
+    public class OnDemandProjectsLoadTests : AbstractMSBuildTestFixture
+    {
+        public OnDemandProjectsLoadTests(ITestOutputHelper output)
+            : base(output)
+        {
+        }
+
+        [Fact]
+        public async Task LoadProjectsOnDemandOneByOne()
+        {
+            var configData = new Dictionary<string, string> { [$"MsBuild:{nameof(MSBuildOptions.OnDemandProjectsLoad)}"] = "true" };
+            using (var testProject = await TestAssets.Instance.GetTestProjectAsync("TwoProjectsWithSolution"))
+            using (var host = CreateMSBuildTestHost(testProject.Directory, configurationData: configData))
+            {
+                MSBuildWorkspaceInfo workspaceInfo = await host.RequestMSBuildWorkspaceInfoAsync();
+
+                // Expect empty workspace initially since no documents have been requested yet
+                Assert.Null(workspaceInfo.SolutionPath);
+                Assert.NotNull(workspaceInfo.Projects);
+                Assert.Equal(0, workspaceInfo.Projects.Count);
+
+                // Requesting library document should load only that project
+                GetCodeActionsService codeActionHandler = host.GetRequestHandler<GetCodeActionsService>(OmniSharpEndpoints.V2.GetCodeActions);
+                GetCodeActionsResponse codeActionResponse = await codeActionHandler.Handle(
+                    new GetCodeActionsRequest { FileName = Path.Combine(testProject.Directory, "Lib", "Class1.cs") });
+                workspaceInfo = await host.RequestMSBuildWorkspaceInfoAsync();
+
+                Assert.NotNull(codeActionResponse);
+                Assert.Null(workspaceInfo.SolutionPath);
+                Assert.NotNull(workspaceInfo.Projects);
+                Assert.Equal(1, workspaceInfo.Projects.Count);
+                Assert.Equal("Lib.csproj", Path.GetFileName(workspaceInfo.Projects[0].Path));
+
+                // Requesting app document should load that project as well
+                QuickFixResponse codeCheckResponse = await host.RequestCodeCheckAsync(Path.Combine(testProject.Directory, "App", "Program.cs"));
+                workspaceInfo = await host.RequestMSBuildWorkspaceInfoAsync();
+
+                Assert.NotNull(codeCheckResponse);
+                Assert.Null(workspaceInfo.SolutionPath);
+                Assert.NotNull(workspaceInfo.Projects);
+                Assert.Equal(2, workspaceInfo.Projects.Count);
+                Assert.Equal("App.csproj", Path.GetFileName(workspaceInfo.Projects[0].Path));
+                Assert.Equal("Lib.csproj", Path.GetFileName(workspaceInfo.Projects[1].Path));
+            }
+        }
+
+        [Fact]
+        public async Task LoadProjectAndItsReferenceOnDemand()
```

Comment:
Worth adding a test for transitives of transitives?

Reply:
@rchande Not sure I'm following. Could you please clarify what test you are proposing to be added?

Reply:
`LoadProjectWithTransitiveReferenceCalledFooAndTransitiveReferencesOfFooAreLoaded`?